lib/salus/scanners/osv/go_osv.rb

```diff
@@ -0,0 +1,128 @@
+require 'salus/scanners/osv/base'
+
+module Salus::Scanners::OSV
+  class GoOSV < Base
+    class SemVersion < Gem::Version; end
+    class SemDependency < Gem::Dependency; end
+
+    EMPTY_STRING = "Not Found".freeze
+    DEFAULT_SOURCE = "https://osv.dev/list".freeze
+    DEFAULT_SEVERITY = "LOW".freeze
+
+    def should_run?
+      @repository.go_sum_present?
+    end
+
+    def run
+      dependencies = find_dependencies
+      if dependencies.empty?
+        msg = "Failed to parse any dependencies from the project."
+        bugsnag_notify("GoOSV: #{msg}")
+        return report_error("GoOSV: #{msg}")
+      end
+
+      if osv_vulnerabilities.nil?
+        msg = "No vulnerabilities found to compare."
+        bugsnag_notify("GoOSV: #{msg}")
+        return report_error("GoOSV: #{msg}")
+      else
+        results = match_vulnerable_dependencies(dependencies)
+      end
+      # Report scanner status
+      if results.length.positive?
+        report_failure
+        log(format_vulns(results))
+      else
+        report_success
+      end
+    end
+
+    # Find dependencies from the project
+    def find_dependencies
+      # Return a map of dependency name and versions used by the project
+      shell_return = run_shell("bin/parse_go_sum #{@repository.go_sum_path}", chdir: nil)
+
+      if !shell_return.success?
+        report_error(shell_return.stderr)
+        return
+      end
+
+      all_dependencies = nil
+      begin
+        all_dependencies = JSON.parse(shell_return.stdout)
+      rescue JSON::ParserError
+        err_msg = "Could not parse JSON returned by bin/parse_go_sum's stdout!"
+        report_stderr(err_msg)
+        report_error(err_msg)
+        return
+      end
+
+      dependencies = {}
+      # Pick specific version of dependencies
+      # If multiple versions of dependencies are found then pick the max version to mimic MVS
+      # https://go.dev/ref/mod#minimal-version-selection
+      all_dependencies.each do |deps|
+        lib = deps["namespace"] + "/" + deps["name"]
+        version = deps["version"].to_s.gsub('v', '').gsub('+incompatible', '')
+        if dependencies.key?(lib)
+          dependencies[lib] = version if SemVersion.new(version) >
+            SemVersion.new(dependencies[lib])
+        else
+          dependencies[lib] = version
+        end
+      end
+      dependencies
+    end
+
+    # Compare vulnerabilities found with dependencies found
+    # and return vulnerable dependencies
+    def match_vulnerable_dependencies(dependencies)
```

Comment:
The method `match_vulnerable_dependencies` has become very long. Is it possible to extract some parts for it to separate smaller methods so that we can make it smaller. 🤔

Reply:
Let me look into it, i was thinking about it but wasn't sure if there was a clean way.

Reply:
Moving version checking to it own function so that should make it much better.